File: cpp/438.find-all-anagrams-in-a-string.cpp
/*
 * @lc app=leetcode id=438 lang=cpp
 *
 * [438] Find All Anagrams in a String
 */

// @lc code=start
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> startIndices;
        if (s.length() < p.length()) {
            return startIndices;
        }

        vector<int> pCharCount(26, 0);
        vector<int> sCharCount(26, 0);

        // correspons to a letter in the alphabet
        // c - 'a' will give the index of the letter in the alphabet
        // by incrementing it, we count the occurence of the letter
        for (char c : p) {
            pCharCount[c - 'a']++;
        }

        // each character s[i] will be converted to an index
        // this is a freq counter for a window from 0 to p.length() - 1
        // we slide this window to the right and increment the occurence
        // of the letter in the window s
        for (int i = 0; i < p.length(); i++) {
            sCharCount[s[i] - 'a']++;
        }

        // if the freq counter of the window is equal to the freq counter of p
        // then we have an anagram
        if (pCharCount == sCharCount) {
            startIndices.push_back(0);
        }


        for (int i = p.length(); i < s.length(); i++) {
            sCharCount[s[i] - 'a']++;
            // calculate the index of the letter that is leaving the window
            // i - position of right end of window
            // p.length - length of the window
            // i - p.length() - index of char that was at the left end
            // sCharCount[...] - convert the char to an index and decrement count

            // exmaple: char is b, then s[i - p.length()] - 'a' will be 1 since b - a = 1
            sCharCount[s[i - p.length()] - 'a']--;

            if (pCharCount == sCharCount) {
                // i - p.length() + 1 will give the starting index of the window
                // example: s = "cbaebabacd", p = "abc"
                // i = 6, p.length() = 3, i - p.length() + 1 = 4
                // the window is "ebab"
                startIndices.push_back(i - p.length() + 1);
            }
        }
        return startIndices;
    }
};
// @lc code=end

File: scala/438.find-all-anagrams-in-a-string.scala
/*
 * @lc app=leetcode id=438 lang=scala
 *
 * [438] Find All Anagrams in a String
 */

// @lc code=start
object Solution {
    def findAnagrams(s: String, p: String): List[Int] = {
        val sLen = s.length
        val pLen = p.length

        if (sLen < pLen) return List()

        val pCharCount = Array.fill(26)(0)
        val sCharCount = Array.fill(26)(0)
        var startIndices = List[Int]()

        for (c <- p) {
            pCharCount(c - 'a') += 1
        }

        for (i <- 0 until pLen) {
            sCharCount(s(i) - 'a') += 1
        }

        if (pCharCount.sameElements(sCharCount)) {
            startIndices = 0 :: startIndices
        }

        for (i <- pLen until sLen) {
            sCharCount(s(i) - 'a') += 1
            sCharCount(s(i - pLen) - 'a') -= 1

            if (pCharCount.sameElements(sCharCount)) {
                startIndices = (i - pLen + 1) :: startIndices
            }
        }

        startIndices.reverse
    }
}
// @lc code=end

